Add Sei blockchain

cordt-sei opened the pull request:
**Summary**:
This revamps handling of Cosmos SDK-based chains and greatly simplifying the addition of new chains. The changes simplify the codebase, reduce redundancy, and make it easier to manage and extend.

### Key Updates:

1. **Centralized Cosmos SDK Logic**:
   - Modified the `FetchCosmosSDKNakaCoeff` function in `cosmos.go` to be exportable and handle all Cosmos SDK-based chains.


2. **New Chain Added**:
   - Added support for the **Sei** network.
   - Updated `chain.go` to include Sei in the list of supported chains.

3. **Streamlined Existing Chains**:
   - Refactored `agoric.go`, `osmosis.go`, `regen.go`, `sei.go`, and `stargaze.go` to use this shared function, cleaning up much redundant code.
   - Included pool data fetching to ensure more accurate Nakamoto coefficient calculations, since the previous method would not have included any for unbonded validators. [correct me if this seems off]

### Testing and Validation:

- **Manual Testing**: Ran tests on each refactored chain to ensure the Nakamoto coefficient calculations are accurate and that everything works as expected.
- **Validation**: Checked logs to confirm the refactored code runs smoothly and handles errors correctly.

### Looking Ahead:

- **Automated Testing**: May be worth adding automated tests for the `FetchCosmosSDKNakaCoeff` function to ensure it continues to work well across different chains. It should work universally as-is, but this may change in the future.
- **Further Simplification**: Could look into whether other non-Cosmos SDK chains could benefit from a similar centralized approach.

Overall this should make the Nakamoto coefficient calculator easier to manage and extend, while keeping all the current features intact.

Here's the image you requested also.

![image](https://github.com/user-attachments/assets/4810b55f-fbab-4909-924c-e5b998790092)

xenowits commented:
@cordt-sei can you please resolve the merge conflicts?

Also, please post a screenshot of the value of nakamoto coefficient that you're getting for Sei.

cordt-sei commented:
> @cordt-sei can you please resolve the merge conflicts?
> 
> Also, please post a screenshot of the value of nakamoto coefficient that you're getting for Sei.

Actually it seems like none of the Cosmos SDK chains are working currently. Is this also what you are seeing?

cordt-sei commented:
@xenowits I've made some pretty large changes, can you please test on your end? It seems to work for me running it directly, but not with docket. I've never really had great luck with docker in general.

xenowits commented:
Thanks for the changes @cordt-sei ! Will review and test it before merging.

xenowits commented on core/chains/sei.go:
```diff
@@ -0,0 +1,7 @@
+func Sei() (int, error) {
+	validatorsURL := "https://rest.sei-apis.com/cosmos/staking/v1beta1/validators?page.offset=1&pagination.limit=100&status=BOND_STATUS_BONDED"
+	poolURL := "https://rest.sei-apis.com/cosmos/staking/v1beta1/pool"
+	return FetchCosmosSDKNakaCoeff("sei", validatorsURL, poolURL)
```
add a newline before the return statement

cordt-sei replied:
Fixed.

xenowits commented on core/chains/cosmos.go:
```diff
@@ -18,70 +18,131 @@ import (
 
-// fetchCosmosSDKNakaCoeff returns the nakamoto coefficient for the provided cosmos SDK-based chain using the provided url.
-func fetchCosmosSDKNakaCoeff(chainName, url string) (int, error) {
+func FetchCosmosSDKNakaCoeff(chainName, validatorURL, poolURL string) (int, error) {
```
add back the comment:
`// fetchCosmosSDKNakaCoeff returns the nakamoto coefficient for the provided cosmos SDK-based chain using the provided url.`

cordt-sei replied:
Re-added.

xenowits commented on core/chains/cosmos.go:
```diff
@@ -93,7 +154,38 @@ func fetch(url string) (cosmosResponse, error) {
+	return response, nil
+}
+
+func fetchPool(url string) (cosmosPoolResponse, error) {
```
add a comment for this function. what's a pool in the context of cosmos chains?

cordt-sei replied:
I'll rename this as well as the validators one for uniformity.

cordt-sei replied:
Further context:

`bonded_tokens` is just the direct way to get all of the valid (consensus/governance weight bearing) staked tokens, which is just a faster way to get the total amount than tallying all of the validators tokens
`not_bonded_tokens` is any that are either currently unstaking, or that are staked to inactive (`not_bonded`) validators.

```json
{
  "pool": {
    "not_bonded_tokens": "8646975757295",
    "bonded_tokens": "359227734437728"
  }
}
```

xenowits commented on core/chains/stargaze.go:
```diff
@@ -1,6 +1,7 @@
-	return fetchCosmosSDKNakaCoeff("stargaze", url)
+	validatorsURL := "https://rest.stargaze-apis.com/cosmos/staking/v1beta1/validators?page.offset=1&pagination.limit=500&status=BOND_STATUS_BONDED"
+	poolURL := "https://rest.stargaze-apis.com/cosmos/staking/v1beta1/pool"
+	return FetchCosmosSDKNakaCoeff("stargaze", validatorsURL, poolURL)
```
add newline before return

cordt-sei replied:
Fixed

xenowits commented on core/chains/cosmos.go:
```diff
@@ -18,70 +18,131 @@ import (
 	Validators []struct {
-		Status string `json:"status"`
-		Tokens string `json:"tokens"`
+		OperatorAddress string `json:"operator_address"`
```
we can probably remove some of these fields in the response as we don't use/need them in our calculations. eg: `Description`, `Commission` etc are probably not required.

cordt-sei replied:
I kept them in incase at any point someone wanted to add some additional detail or context but yeah it's bloat anyway.

xenowits reviewed (approved):
LGTM 🚀

xenowits commented:
@cordt-sei thanks for the cleanup :)

cordt-sei commented:
> @cordt-sei thanks for the cleanup :)

In addition to addressing the review comments, I added Juno chain to the standardized Cosmos chains and cleaned up some other minor things..

xenowits commented on core/chains/cosmos.go:
```diff
@@ -18,82 +18,154 @@ import (
 }
-func fetch(url string) (cosmosResponse, error) {
+
+// Fetches
```
please complete this comment.

cordt-sei replied:
added

xenowits commented on core/chains/cosmos.go:
```diff
@@ -18,82 +18,154 @@ import (
+	return response, nil
+}
+
+func fetchStakingPoolData(url string) (cosmosStakingPoolData, error) {
```
also add a comment for this function

cordt-sei replied:
added

xenowits commented:
@cordt-sei I have tested this PR. Please fix the above comments and then I can merge it.
